### ISSUE TITLE
Segfault when using schema evolution

### ISSUE THREAD

**Issue by pcanal** [root-project/root#7754]
See https://root-forum.cern.ch/t/segfault-when-using-schema-evolution/43765

Hello,

We have an issue that we have not been able to solve, nor understand. We hope you can provide some help.

We have a class written into a TTree with one of its Base class having been modified. We have udpated the class version during the modification. In the ROOT file, we have version 1 of the class (TVHit) while in the current software, we have version 2 of the class. We wrote in the LinkDef file a #pragma instruction on how to transform v1 into v2. When this #pragma line is omitted, we obviously read the wrong information from the input file, because the definition in the code does not correspond to the definition on file (as expected). But when using the #pragma line, we simply get a segfault while trying to read the TTree.

I have prepared a self-contained minimal working example attached to this topic: [mwe.tgz|attachment](upload://34Fdg60DY7tGGh6cw40YICFFhLy.tgz) (224.3 KB) 

The archive contains the following README.md with more details on how to reproduce the issue, and on the exact content of the archive:

Steps to reproduce issue on Minimal Working Example (MWE)
=============================
  - Set the ROOT_INCLUDE_PATH and LD_LIBRARY_PATH
    `source env.sh`
  - Compile the minimal persistency library
    `make`
  - Run test code on new persistency (opens pluto_newpers.root and reads the tree successfully)
    `root read_updated.C\(1\)`
    or
    `read_updated 1`
  - Run test code on old persistency (opens pluto_oldpers.root and crashes when trying to read the tree)
    `root read_updated.C(\(0\)`
    or
    `read_updated 0`

The crash is traced back to line 109 of PersistencyDICT.cxx
`static Long_t offset_Onfile_TVHit_fMCTrackID = oldObj->GetClass()->GetDataMemberOffset("fMCTrackID");`
where GetClass returns a nullptr.

Content of the archive
======================
  - The header files in include contain the current definition of the minimal persistency necessary to 
    read the input files (TVHit version 2).
  - The source files in src are there for the sake of compilation but should not
    have an influence on reading the input file (right?). 
  - The header files in include_oldpers are included for documentation and correspond to the difference
    between the old persistency (TVHit version 1) and the new persistency (TVHit version 2).
  - PersistencyLinkDef.hh is the current linkdef file. The "#pragma" line at the end is our attempts
    to make the previous persistency compatible with the current one. Essentially the field
    `fMCTrackID` is replaced with the field `fKinePartIndex` and the field `fMCTrackID` still
    exists, but with a different purpose. So `fMCTrackID` in version 1 should be forwarded to
    `fKinePartIndex` in version 2.
  - pluto_oldpers.root is the input ROOT file containing a tree with 10 entries using TVHit version 1
  - pluto_newpers.root is the input ROOT file containing a tree with 10 entries using TVHit version 2

___
_ROOT Version: 6.22/00
_Platform: CentOS7 (lxplus)
_Compiler: gcc 8.3.0 (LCG_98python3)
___

**Comment by nlurkin** [comment]
Thank you! Do I understand correctly that this will be part of a patch to 6.22 ?

**Comment by pcanal** [comment]
Yes, the fix is already in the v6.22 patch branch and will part of the v6.22/10 releases.

**Comment by nlurkin** [comment]
Thank you, what is the planned timeline for that release?

**Comment by Axel-Naumann** [comment]
Issue exists in master while PR triggering this was reverted in 6.24

**Comment by nlurkin** [comment]
Hello,
We were trying to migrate some of our production to LCG100 today and we were surprised to discover that we get the crash again with 6.24. 
Does this above means that the patch was reverted and is not actually present in any existing ROOT release?

**Comment by pcanal** [comment]
Unfortunately the patch was not complete and broke other use case, so it has been reverted in v6.24 while we work on it :(

**Comment by nlurkin** [comment]
Hello, what is the status for this?
Is it going to be included in any 6.24 patch soon? We currently have a workaround so that our current software works, but we have some necessary ongoing development that is incompatible with the workaround and needs the patch for this issue. This necessary development has been on hold for several months already.

**Comment by Axel-Naumann** [comment]
@pcanal can you share your plans / ETA / an update?

**Comment by nlurkin** [comment]
@pcanal Any update please? The work on our side that depends on this has been pending for many months already.

**Comment by pcanal** [comment]
Still no ETA but actively working on it.  I got to the point where I can reproduce and analyze the negative side effect of the patch.

**Comment by pcanal** [comment]
Note: I am confused :( ... I had been able to use your example and solve the problem (in a way that turns out to not be sufficient) but now that I return to it (to make sure the 'new way' is also working), I am having some "unrelated" problem.

With any release of ROOT  i tried (including v6.22/00) and NONE of your libraries around, I get:
```
root.exe -b -l -q pluto_oldpers.root -e 'b = SlimMC->GetBranch("fKineParts.fFinalMomentum"); b->GetEntry(0);'

Attaching file pluto_oldpers.root as _file0...
Warning in <TClass::Init>: no dictionary for class Stream is available
Warning in <TClass::Init>: no dictionary for class RecoInfo is available
Warning in <TClass::Init>: no dictionary for class AnalysisInfo is available
Warning in <TClass::Init>: no dictionary for class CheckPointInfo is available
Warning in <TClass::Init>: no dictionary for class NA62Analysis::Core::AnalyzerIdentifier is available
Warning in <TClass::Init>: no dictionary for class TDetectorVEvent is available
Warning in <TClass::Init>: no dictionary for class TVEvent is available
Warning in <TClass::Init>: no dictionary for class TDetectorVHit is available
Warning in <TClass::Init>: no dictionary for class TVHit is available
Warning in <TClass::Init>: no dictionary for class TVChannelID is available
Warning in <TClass::Init>: no dictionary for class TSpectrometerEvent is available
Warning in <TClass::Init>: no dictionary for class TSpectrometerHit is available
Warning in <TClass::Init>: no dictionary for class SpectrometerChannelID is available
Warning in <TClass::Init>: no dictionary for class SlimEvent is available
Warning in <TClass::Init>: no dictionary for class SlimGenePart is available
Warning in <TClass::Init>: no dictionary for class SlimBasePart is available
Warning in <TClass::Init>: no dictionary for class SlimKinePart is available
Warning in <TClass::Init>: no dictionary for class Float4Vector is available
Warning in <TClass::Init>: no dictionary for class Float3Vector is available
Warning in <TClass::Init>: no dictionary for class EventBoundary is available
Warning in <TClass::Init>: no dictionary for class MCInfo is available
(TFile *) 0x55f64bf49770
Error in <TBufferFile::ReadClassBuffer>: class: TVector3, attempting to access a wrong version: 1348, object skipped at offset 96
Error in <TBufferFile::ReadClassBuffer>: class: TVector3, attempting to access a wrong version: 3184, object skipped at offset 102
Error in <TBufferFile::CheckByteCount>: object of class TVector3 read too few bytes: 2 instead of 627270009
Error in <TBufferFile::CheckByteCount>: Byte count probably corrupted around buffer position 96:
        627270009 for a possible maximum of 35226
Error in <TBufferFile::CheckByteCount>: object of class TVector3 read too many bytes: 40 instead of -1073741825
Warning in <TBufferFile::CheckByteCount>: TVector3::Streamer() not in sync with data on file pluto_oldpers.root, fix Streamer()
Error in <TBufferFile::CheckByteCount>: Byte count probably corrupted around buffer position 35328:
        -1073741825 for a possible maximum of -44
Error in <TBufferFile::CheckByteCount>: object of class TVector3 read too many bytes: 40 instead of -1073741825
Warning in <TBufferFile::CheckByteCount>: TVector3::Streamer() not in sync with data on file pluto_oldpers.root, fix Streamer()
Error in <TBufferFile::CheckByteCount>: Byte count probably corrupted around buffer position 35328:
        -1073741825 for a possible maximum of -44
Error in <TBufferFile::CheckByteCount>: object of class TVector3 read too many bytes: 40 instead of -1073741825
Warning in <TBufferFile::CheckByteCount>: TVector3::Streamer() not in sync with data on file pluto_oldpers.root, fix Streamer()
Error in <TBufferFile::CheckByteCount>: Byte count probably corrupted around buffer position 35328:
        -1073741825 for a possible maximum of -44
Error in <TBufferFile::CheckByteCount>: object of class TVector3 read too many bytes: 40 instead of -1073741825
Warning in <TBufferFile::CheckByteCount>: TVector3::Streamer() not in sync with data on file pluto_oldpers.root, fix Streamer()
Error in <TBufferFile::CheckByteCount>: Byte count probably corrupted around buffer position 35328:
        -1073741825 for a possible maximum of -44
```

where the warning are expected but the Error are not expected (and I don't recall seeing them).

Do you have a similar problem or is it clean in your case?

**Comment by nlurkin** [comment]
Hello @pcanal 
I confirm that using the command you provide, I get the same crash.

**Comment by pcanal** [comment]
Do you have other similar files (and/or can you still generate that type of file)?  If you do, do they also have the same problem (i.e. wondering if that file is somehow corrupted or if something is going on.  eg. the 'new' file does not have that problem).

**Comment by nlurkin** [comment]
I will have to check. I do not remember now how I generated those files.

**Comment by nlurkin** [comment]
I think I managed to produce a new file the same way as I did in the past. You can find it at
`/afs/cern.ch/user/n/nlurkin/public/pluto_oldpers.root`
However it has the same issue as the one you have (probably due to filtering and skimming using a software that has the new persistency)
```
Info in <TBranchElement::InitializeOffsets>: TTree created with an older schema, some data might not be copied in 'slow-cloning' mode; fast-cloning should have the correct result. 'fDirectParent' is missing when constructing the branch 'fKineParts'. 
Info in <TBranchElement::InitializeOffsets>: TTree created with an older schema, some data might not be copied in 'slow-cloning' mode; fast-cloning should have the correct result. 'fID' is missing when constructing the branch 'fKineParts'. 
Info in <TBranchElement::InitializeOffsets>: TTree created with an older schema, some data might not be copied in 'slow-cloning' mode; fast-cloning should have the correct result. 'fParentID' is missing when constructing the branch 'fKineParts'. 
```
The original file (which doesn't show this particular issue you spotted, but still the one that we want to solve) can be found at (but is much bigger)
`/afs/cern.ch/user/n/nlurkin/public/pluto_c123_dr008094_r4420812.root`

**Comment by pcanal** [comment]
> probably due to filtering and skimming using a software that has the new persistency

CloneTree currently can only handle schema evolution if the 'fast' option is selected.  For slow cloning it does not (yet?) deal well with the fact that the on-file and in-memory representation does not match (the reading part is 'sorta' fine but the writing part is confused because the cloned branches are (for now) the exact same as the input and thus we are now in the case were we attempt to write in a different format than the one in memory.   The "Info" above are clear/exact that some data is lost because those field are no longer in memory (but they still have a branch for them on the output, so behavior will be strange).

I am unclear why this leads to the behavior I have seen/reported but it might be linked.

Thanks to your new file (that I can clone to a smaller if I make sure to load none of the pluto libraries), I seem to be back on track.  thanks.

**Comment by pcanal** [comment]
Note that this fix will be part of the next ROOT release (v6.26/00 coming soon) [and of course be back ported as needed].
